firestore/src/main/java/com/firebase/ui/firestore/paging/QueryFactory.kt

```diff
@@ -0,0 +1,261 @@
+package com.firebase.ui.firestore.paging
+
+import com.google.firebase.firestore.CollectionReference
+import com.google.firebase.firestore.FieldPath
+import com.google.firebase.firestore.Query
+import com.google.firebase.firestore.Query.Direction.*
+
+/**
+ * Factory for direct and reverse [Query].
+ *
+ * @property collectionReference The collection to query.
+ * @property filters The filters to apply.
+ */
+class QueryFactory private constructor(
+        private val collectionReference: CollectionReference,
+        private val filters: List<Filter>
+) {
+    /**
+     * Creates and returns a new direct [Query] that's filtered by the specified filters.
+     *
+     * @return The created [Query].
+     */
+    fun create(): Query = filters.fold(collectionReference as Query) { query, filter -> filter.apply(query) }
+
+    /**
+     * Creates and returns a new reverse [Query] that's filtered by the specified filters.
+     *
+     * @return The created [Query].
+     */
+    fun createReverse(): Query =
+            filters.fold(collectionReference as Query) { query, filter -> filter.applyReverse(query) }
+
+    /**
+     * Base class for filters.
+     *
+     * @property fieldPath The field to sort by.
+     */
+    private abstract class Filter(
```

Comment:
One of the informal rules we have on the team is "don't re-implement the query interface".  The reason is that this library will always be chasing the official SDK as queries are added / changed.  Plus as you can see by your implementation, it's a lot of code.

From what I can tell this system is used to be able to create a reversible query.  Have you considered just asking the developer to pass in both a forward and reverse Query object (with no `limit()`, `startAt()` or `endAt()` clauses) when they create the recycler adapter?

Reply:
Yep, we can ask for backward query, but as I understand only a few people understand that to query documents before provided you should create reverced query, not just call endAt + limit. 
To tell the truth I myself find out this only while working on this feature. =)

Reply:
Yeah, this whole file should probably be removed. 🙂